### ISSUE TITLE
Broken AT Command Interface

### ISSUE THREAD

**Issue by SecUpwN** [SecUpwN/Android-IMSI-Catcher-Detector#754]
I am always using the latest version compiled from our `development` branch and my [testing device](https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/wiki/Testing-Devices) - it seems as if the `AT Command Interface` does not react to any command on `/dev/smd0`, especially not those described in the article [Get Kc key and TMSI number](https://ferrancasanovas.wordpress.com/2014/01/28/get-kc-key-and-tmsi-number/). @larsgrefer, are you able to confirm this?

**Comment by ecs87** [comment]
Hey there,

I've been lurking around for a while looking at what you guys have been doing and decided to give the APK a try. I also couldn't get the AT command interface to work on my G920P on /dev/smd0. I didn't really expect it to work, seeing as how I can't get any terminal to get a response from the phone on any of the /dev ports. It however still communicate over a UART interface when I dial *#0*#.

I know I'm VERY late to the party but I also have some goodies that you guys might be interested in. I'm not sure if you guys are familiar with GSM boxes, but some of them have developed their own binary that they push to /data/local/tmp and run that can send AT commands to the phone (even UART level AT commands such as AT+AKSEEDNO, AT+MSLSECUR, etc). Their earlier binary needed root, however their newer one does not. The newer one however does have some kind of security on it, which I have no clue how it works. I was able to see it send an AT+INIT command to start the program, and I assume this was the security on it.

Here's the earlier binary I mentioned:

[AT.zip](https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/files/99967/AT.zip)

Here's the later binary that does not need root:

[NewATBinaries.zip](https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/files/99968/NewATBinaries.zip)

Hope this helps you guys. I'll be checking in to answer any questions that may come up. 

-ecs87

**Comment by SecUpwN** [comment]
>I know I'm VERY late to the party but I also have some goodies that you guys might be interested in.

Huge thanks for the binary and welcome to our little party, @ecs87! I am not sure yet how and if we can integrate this, but I am confident @0x7678 will hopefully have a closer look at this since he's our expert.

**Comment by ecs87** [comment]
No problemo. Well, you guys have been interested in reading information from the RIL, which I assume this program does. It has DIRECT access to RIL or maybe the modem; I'm not entirely sure seeing as how I'm not a baseband engineer/programmer (more on the web-programming end of things, but reverse engineering the baseband is a hobby of mine), but the AT commands I've seen it send aren't recognized over the regular modem port it makes with the PC in QC DM mode, and NEITHER over UART (increased access to more AT commands). This is the only reason I believe that it has higher access; either that or it sends IPC commands that they've created AT commands for (your AT command through their binary > AT command turns into an IPC message > RIL > MODEM > Possibly writes to the encrypted EFS data partitions {I don't mean EFS; I mean modemst1, modemst2, FSG in the terms of a Qualcomm baseband chipset. Exynos would be the only one that would write to the EFS partition to the nv_data.bin file...AND I MEAN AN EXYNOS CP, not a AP. The G920P has an Exynos AP but a Qualcomm CP meaning it DOES NOT use /efs/ to store any significant baseband data}).

Whew...that was a huge wall of text lol. Sorry about that.

**Comment by 0x7678** [comment]
@SecUpwN and @ecs87 thank you for this, let me look at them during the week, superb input by the way.

**Comment by banjaxbanjo** [comment]
@ecs87 Very interesting stuff and didn't know binaries like this existed.

dev/smd0 isn't always going to process the AT commands and you should do process of elimination of all the smd ports to see if others are more friendly at giving back data.

On mtk phones atci (name similar to that ) ports are good for AT commands think ports are in dev/radio or something.

My Gaxlaxy S5 smd 0 7 9 are good and on Sony xperia 7 11 are good.

Sony Xperia Its possible to get tmsi and kc from AT commands but galaxy S5 gives back nothing with same commands.

Connect your phone to adb shell with su and try echo -e "AT\r" > dev/smd0 and try other ports if any smd1 2 3 4 5 6 etc.. you should get an OK if success.

Open another shell at same time:
logcat -b radio to see if any response from AT command or cat dev/smd0 <---- change this to the port you are checking at the time smd1 2 3 etc... 

My memory is a bit hazy in the cat command but you get what I mean.

**Comment by ecs87** [comment]
You're most welcome. These files came out of my z3x installation directory. I'm positive that they would never release the source of any files, so unfortunately the reversing process might be a huge pain in the ass. Sorry in advance!

Also another addition would be what I think is their AT RIL binary for the SHANNONxxx chipsets (including the SHANNON333, their most recent Exynos baseband chipset in the S6, S6 Edge, Note 5, and S6 Edge plus).

[zTool.zip](https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/files/102475/zTool.zip)

**Comment by banjaxbanjo** [comment]
@0x7678 awesome blog by the way some amazing stuff on there and great source of information👍.

**Comment by 0x7678** [comment]
@ecs87 its quite OK, will definnatley look at them and how they will reflect on the project, also had issues with the AT cok=mmands on my nexus 6, will retry after pushing some of this :)

@banjaxbanjo thank you, really humbled, been silent on it though... will add more content :)

**Comment by SecUpwN** [comment]
@tt3mm, since you removed the `ATCommandActivity` with https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/commit/9478d1ee23c2967e4d6a866aa6faba6cddf05ef5, is this Interface still functional?

**Comment by tt3mm** [comment]
Yes, don't worry, the same code lives in a form of a fragment :)
At least for me it works as before.

I have 2 SIM cards, and I was trying to make it detect both, but I couldn't. At least I worked in a shameful way, of trying AT-ing all /dev/smd*.
Any ideas how can it be fixed?

**Comment by SecUpwN** [comment]
>I have 2 SIM cards, and I was trying to make it detect both, but I couldn't. At least I worked in a shameful way, of trying AT-ing all /dev/smd*. Any ideas how can it be fixed?

@tt3mm, seems like we lack support for two SIM cards as of yet. Have a look at #504.

**Comment by webmip** [comment]
Hi All,
The same problem with the AT commands & nexus 5, I will try to send the commands via adb, but don't get the OK success.

**Comment by ecs87** [comment]
@SecUpwN does /system/bin/at_distributor do anything to help this bug? When the software I use for GSM phones initializes "AT RIL" (the AT binary I provided) it uses this command through the adb shell "setprop ctl.stop at_distributor". Hmmmm...possibly they've hacked at_distributor and secured it (using an +INIT command)? The at_distributor file should give you an interface to AT commands at the highest level (minus using the command DEVROOTK, but the rest of the UART commands seem to work). I haven't quite gotten it to work though...maybe you'll have more luck.

**Comment by aseel-alking** [comment]
> مرحبا،
> 
> لقد كنت أتجول لفترة من الوقت وأنا أنظر إلى ما كنت تفعله يا رفاق وقررت أن أجرب APK. لم أتمكن أيضًا من تشغيل واجهة أوامر AT على G920P على / dev / smd0. لم أكن أتوقع حقًا أن يعمل ، لأنني رأيت كيف لا يمكنني الحصول على أي محطة طرفية للحصول على استجابة من الهاتف على أي من منافذ / dev. ومع ذلك ، فإنه لا يزال يتواصل عبر واجهة UART عندما أطلب (علامة النجمة) # 0 (علامة النجمة) #.
> 
> أعلم أنني تأخرت كثيرًا في الحفلة ولكن لدي أيضًا بعض الأشياء الجيدة التي قد تكون مهتمًا بها يا رفاق. لست متأكدًا مما إذا كنتم على دراية بصناديق GSM ، لكن البعض منهم طور ثنائيته الخاصة التي يدفعونها إلى / data / local / tmp وتشغيلها التي يمكنها إرسال أوامر AT إلى الهاتف (حتى أوامر UART على مستوى AT مثل AT + AKSEEDNO ، AT + MSLSECUR ، إلخ). يحتاج النظام الثنائي السابق إلى الجذر ، ولكن لا يحتاج الجذر الأحدث. ومع ذلك ، يحتوي الإصدار الأحدث على نوع من الأمان عليه ، وليس لدي أدنى فكرة عن كيفية عمله. لقد تمكنت من رؤيته يرسل أمر AT + INIT لبدء البرنامج ، وأفترض أن هذا كان الأمان عليه.
> 
> هذا هو الثنائي السابق الذي ذكرته:
> 
> [AT.zip](https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/files/99967/AT.zip)
> 
> إليك الخيار الثنائي الأخير الذي لا يحتاج إلى الجذر:
> 
> [NewATBinaries.zip](https://github.com/SecUpwN/Android-IMSI-Catcher-Detector/files/99968/NewATBinaries.zip)
> 
> آمل أن يساعد هذا يا رفاق. سأقوم بالتحقق للإجابة على أي أسئلة قد تطرأ.
> 
> -ecs87

plzzz how to warkng plzzz
